src/rotatez.jl

```diff
@@ -506,3 +498,78 @@ function imrotate_adj!(output::AbstractMatrix{<:RealU},
     end
     return output
 end
+
+
+# prepare for "foreach" threaded computation
+_setup = (z) -> Channel() do ch
+    foreach(i -> put!(ch, i), z)
+end
+
+
+"""
+    imrotate!(output, image3, θ, plans)
+In-place version of rotating a 3D `image3` by `θ`
+in counter-clockwise direction (opposite to `imrotate` in Julia)
+"""
+function imrotate!(
+    output::AbstractArray{<:RealU,3},
+    image3::AbstractArray{<:RealU,3},
+    θ::RealU,
+    plans::Vector{<:PlanRotate},
+)
+
+    size(output) == size(image3) || throw(DimensionMismatch())
+
+    fun = z -> imrotate!(
+        (@view output[:,:,z]),
+        (@view image3[:,:,z]),
+        θ,
+        plans[Threads.threadid()],
+    )
+
+    ntasks = length(plans)
+    Threads.foreach(fun, _setup(1:size(image3,3)); ntasks)
+
+#=
+    Threads.@threads for z = 1:size(image3,3) # 1:nz
+        id = Threads.threadid() # thread id
+
+        imrotate!(
+            (@view output[:,:,z]),
+            (@view image3[:,:,z]),
+            θ,
+            plans[id],
+        )
+    end
+=#
+
+    return output
+end
+
+
+"""
+    imrotate_adj!(output, image3, θ, plan)
+In-place version of the adjoint of rotating a 3D `image3` by `θ`
+in counter-clockwise direction (opposite to `imrotate` in Julia)
+"""
+function imrotate_adj!(
+    output::AbstractArray{<:RealU,3},
+    image3::AbstractArray{<:RealU,3},
+    θ::RealU,
+    plans::Vector{<:PlanRotate},
+)
+
+    size(output) == size(image3) || throw(DimensionMismatch())
+
+    fun = z -> imrotate_adj!(
+        (@view output[:,:,z]),
+        (@view image3[:,:,z]),
+        θ,
+        plans[Threads.threadid()],
+    )
+
+    ntasks = length(plans)
+    Threads.foreach(fun, _setup(1:size(image3,3)); ntasks)
```

Comment:
I see, so we might want to use `foreach` everywhere?

Reply:
Probably - after we do timing tests of this...